File: problems/451/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var frequencySort = function(s) {
    const m = {};
    
    for (let ch of s) {
        if (!m[ch]) {
            m[ch] = 0;
        }
        
        m[ch]++;
    }
    
    const buckets = new Array(Object.keys(m).length + 1);
    
    for (let key of Object.keys(m)) {
        const val = m[key];
        
        if (!buckets[val]) {
            buckets[val] = [];
        }
        
        buckets[val].push(key);
    }
    
    let ans = '';
    
    for (let i = buckets.length - 1; i >= 0 ; i--) {
        if (!buckets[i] || !buckets[i].length) {
            continue;
        }
        
        for (let ch of buckets[i]) {
            ans += ch.repeat(i);
        }
    }
    
    return ans;
};

File: problems/451/solution.cpp
class Solution {
public:
    string frequencySort(string s) {
        string ans;
        unordered_map<char, int> freqs;
        vector<pair<char, int>> freq;
        
        for (const char& ch: s) {
            freqs[ch]++;
        }
        
        for (auto [ch, cnt]: freqs) {
            freq.push_back({ch, cnt});
        }
        
        sort(begin(freq), end(freq), [](const auto& a, const auto& b) {
            if (a.second == b.second) {
                return a.first < b.first;
            }
            
            return a.second > b.second;
        });
        
        for (auto [ch, cnt]: freq) {
            ans += string(cnt, ch);
        }
        
        return ans;
    }
};